File: src/main/java/top/chenray/antilitematica/punish/GraduatedPunisher.java
package top.chenray.antilitematica.punish;

import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import top.chenray.antilitematica.AntiLitematicaPlugin;
import top.chenray.antilitematica.config.Settings;
import top.chenray.antilitematica.punish.hook.AdvancedBanHook;
import top.chenray.antilitematica.punish.hook.BanPluginHook;
import top.chenray.antilitematica.punish.hook.EssentialsHook;
import top.chenray.antilitematica.punish.hook.LiteBansHook;
import top.chenray.antilitematica.punish.hook.NoOpBanHook;
import top.chenray.antilitematica.util.DiscordWebhook;
import top.chenray.antilitematica.util.Msg;

/**
 * Graduated punishment system: escalates penalties based on violation count within a time window.
 */
public final class GraduatedPunisher {
   private final AntiLitematicaPlugin plugin;
   private final Settings settings;
   private final PunishmentTracker tracker;
   private BanPluginHook hook;
   private final Map<UUID, Integer> notifiedLevel = new ConcurrentHashMap<>();

   public GraduatedPunisher(AntiLitematicaPlugin plugin, Settings settings, PunishmentTracker tracker) {
      this.plugin = plugin;
      this.settings = settings;
      this.tracker = tracker;
      this.resolveHook();
   }

   private void resolveHook() {
      Settings.GraduatedPunishment gp = this.settings.graduatedPunishment();
      if (gp == null || !gp.enabled()) {
         this.hook = new NoOpBanHook();
         return;
      }
      for (String name : gp.banPlugins()) {
         BanPluginHook candidate = createHook(name);
         if (candidate != null && candidate.isAvailable()) {
            this.hook = candidate;
            this.plugin.getLogger().info("Graduated punishment using ban plugin: " + candidate.getName());
            return;
         }
      }
      this.hook = new NoOpBanHook();
      this.plugin.getLogger().info("Graduated punishment using Bukkit native ban/kick.");
   }

   private static BanPluginHook createHook(String name) {
      String lower = name.toLowerCase(Locale.ROOT);
      if (lower.contains("litebans")) {
         return new LiteBansHook();
      } else if (lower.contains("advancedban")) {
         return new AdvancedBanHook();
      } else if (lower.contains("essentials")) {
         return new EssentialsHook();
      }
      return null;
   }

   public void punish(Player player, String channel, String why) {
      Settings.GraduatedPunishment gp = this.settings.graduatedPunishment();
      if (gp == null || !gp.enabled()) {
         return;
      }

      ViolationRecord record = this.tracker.recordViolation(player);
      int level = Math.min(record.count(), gp.levels().size());
      boolean exceeded = record.count() > gp.levels().size();

      Settings.PunishmentLevel pl = exceeded ? gp.exceedMax() : gp.levels().get(level - 1);
      if (pl == null) {
         pl = gp.exceedMax();
      }

      String reason = pl.reason()
            .replace("%player%", player.getName())
            .replace("%count%", String.valueOf(record.count()))
            .replace("%total%", String.valueOf(record.totalViolations()));

      String actionName = pl.action().toUpperCase(Locale.ROOT);
      this.plugin.getLogger().info("[GraduatedPunish] " + player.getName() + " level=" + level +
            " count=" + record.count() + " action=" + actionName + " reason=" + reason);

      // Execute action
      switch (pl.action().toLowerCase(Locale.ROOT)) {
         case "warn":
            this.hook.warn(player, Msg.color("&e[Warning] &f" + reason));
            break;
         case "kick":
            this.hook.kick(player, Msg.color(reason));
            break;
         case "tempban":
            long seconds = parseDuration(pl.duration());
            this.hook.tempBan(player, Msg.color(reason), seconds);
            break;
         case "ban":
            this.hook.ban(player, Msg.color(reason));
            break;
         default:
            this.plugin.getLogger().warning("Unknown punishment action: " + pl.action());
            break;
      }

      // Broadcast
      if (pl.broadcast()) {
         String msg = Msg.color("&c[AntiLitematica] &e" + player.getName() + " &7was punished (&f" +
               actionName + "&7) for using projection mods. &8(" + record.count() + " violations)");
         Bukkit.broadcast(msg, "bukkit.broadcast.user");
      }

      // Staff alert
      if (pl.staffAlert()) {
         String alert = Msg.color("&c[AntiLitematica] &e" + player.getName() + " &7punished (&f" +
               actionName + "&7): &f" + reason + " &8[count=" + record.count() + ",total=" + record.totalViolations() + "]");
         for (Player online : Bukkit.getOnlinePlayers()) {
            if (online.hasPermission("antilitematica.notify")) {
               online.sendMessage(alert);
            }
         }
      }

      // Integration flag
      if (this.settings.integration().enabled()) {
         Settings.Integration integ = this.settings.integration();
         this.plugin.getIntegrationManager().flag(player, integ.checkPrefix() + ":graduated_" + pl.action(),
               integ.violationLevel(), "graduated punishment level=" + level + " action=" + actionName);
      }

      // Discord notification
      sendDiscordNotification(player, channel, why, actionName, reason);
   }

   public void reload() {
      this.resolveHook();
      this.notifiedLevel.clear();
   }

   private long parseDuration(String duration) {
      if (duration == null || duration.isEmpty() || duration.equals("0")) {
         return 0;
      }
      duration = duration.trim().toLowerCase(Locale.ROOT);
      try {
         if (duration.endsWith("d")) {
            return Long.parseLong(duration.substring(0, duration.length() - 1)) * 86400L;
         } else if (duration.endsWith("h")) {
            return Long.parseLong(duration.substring(0, duration.length() - 1)) * 3600L;
         } else if (duration.endsWith("m")) {
            return Long.parseLong(duration.substring(0, duration.length() - 1)) * 60L;
         } else if (duration.endsWith("s")) {
            return Long.parseLong(duration.substring(0, duration.length() - 1));
         } else {
            return Long.parseLong(duration);
         }
      } catch (NumberFormatException e) {
         this.plugin.getLogger().warning("Invalid duration format: " + duration);
         return 0;
      }
   }

   private void sendDiscordNotification(Player player, String channel, String why, String action, String reason) {
      Settings.Discord dc = this.settings.discord();
      if (dc != null && dc.enabled() && dc.webhookUrl() != null && !dc.webhookUrl().isEmpty()) {
         if (dc.notifyOnPunish()) {
            DiscordWebhook webhook = new DiscordWebhook(
                  this.plugin,
                  dc.webhookUrl(),
                  dc.username(),
                  dc.avatarUrl() != null ? dc.avatarUrl() : "",
                  dc.embedTitle(),
                  dc.embedColor(),
                  dc.footerText() != null ? dc.footerText() : "",
                  dc.proxyHost() != null ? dc.proxyHost() : "",
                  dc.proxyPort(),
                  dc.proxyUsername() != null ? dc.proxyUsername() : "",
                  dc.proxyPassword() != null ? dc.proxyPassword() : ""
            );
            webhook.sendDetection(player.getName(), player.getUniqueId().toString(), channel,
                  reason != null ? reason : "Unknown", action);
         }
      }
   }
}

File: src/main/kotlin/icu/epochcraft/antilitematica/punish/GraduatedPunisher.kt
package icu.epochcraft.antilitematica.punish

import icu.epochcraft.antilitematica.AntiLitematica
import icu.epochcraft.antilitematica.database.BanRecord
import icu.epochcraft.antilitematica.database.DetectionRecord
import icu.epochcraft.antilitematica.detection.ActionType
import icu.epochcraft.antilitematica.detection.ChannelRegistry
import icu.epochcraft.antilitematica.detection.DetectionContext
import icu.epochcraft.antilitematica.detection.DetectionHandler
import icu.epochcraft.antilitematica.event.DetectionType
import icu.epochcraft.antilitematica.util.MessageUtil
import org.bukkit.entity.Player

/**
 * 渐进惩罚：按违规次数逐级升级（WARN → KICK → TEMPBAN → BAN）。
 *
 * 作为 [DetectionHandler] 注册到检测总线，认领所有检测并执行升级惩罚。
 *
 * 相对旧版 Java 版修复：
 *   - 等级表为空时不再抛 IndexOutOfBoundsException（空表防御）
 *   - 支持 antilitematica.bypass 权限豁免
 *   - 全部走统一检测总线（不再有独立的第三条惩罚路径）
 *
 * @author 阿清
 */
class GraduatedPunisher(
    private val plugin: AntiLitematica,
    private val tracker: ViolationTracker,
) : DetectionHandler {

    override fun handle(ctx: DetectionContext): Boolean {
        // 渐进惩罚未开启 → 放行给 DetectionPunisher（基础动作）
        if (!plugin.configHolder.graduatedPunishment.enabled) return false
        // 环境通道（fml:/fabric:）→ 放行（强制 LOG，防误伤 Forge/Fabric）
        if (ctx.detectionType == DetectionType.CHANNEL && ChannelRegistry.isEnvironmentChannel(ctx.channel)) return false
        punish(ctx.player, ctx.channel, ctx.reason)
        return true
    }

    /** 执行一次渐进惩罚 */
    fun punish(player: Player, channel: String, reason: String) {
        val cfg = plugin.configHolder.graduatedPunishment
        if (!cfg.enabled) return

        // 权限豁免（测试号 / 白名单）
        if (player.hasPermission("antilitematica.bypass")) return

        val record = tracker.recordViolation(player, player.world.name)
        val levels = cfg.levels

        // 空等级表防御（旧版 B3 bug 修复）
        if (levels.isEmpty()) {
            plugin.logger.warning("渐进惩罚等级表为空，已跳过惩罚（请检查 config.yml graduated-punishment.levels）")
            return
        }

        // 当前级别：按次数取 level，超限使用 exceedMax
        val level: PunishmentLevel =
            if (record.count > levels.size) cfg.exceedMax
            else levels[record.count - 1]

        val msg = level.reason
            .replace("{player}", player.name)
            .replace("{count}", record.count.toString())
            .replace("{total}", record.totalViolations.toString())
            .let { MessageUtil.colorize(it) }

        plugin.logger.info(
            "[GraduatedPunish] ${player.name} count=${record.count} " +
                "action=${level.action.name} reason=$msg"
        )

        // ---- 执行动作 ----
        when (level.action) {
            PunishmentAction.WARN -> player.sendMessage(msg)
            PunishmentAction.KICK -> plugin.server.scheduler.runTask(plugin, Runnable {
                if (player.isOnline) player.kickPlayer(msg)
            })
            PunishmentAction.TEMPBAN ->
                plugin.banManager.ban(player.uniqueId, player.name, msg, level.durationMillis)
            PunishmentAction.BAN ->
                plugin.banManager.ban(player.uniqueId, player.name, msg, BanRecord.PERMANENT)
        }

        // ---- 全服广播 ----
        if (level.broadcast) {
            val broadcast = MessageUtil.colorize(
                "&c[AntiLitematica] &e${player.name} &7因使用投影 mod 被处理 (&f${level.action.displayName}&7) &8(第 ${record.count} 次)",
            )
            plugin.server.onlinePlayers.forEach { it.sendMessage(broadcast) }
        }

        // ---- 管理员通知 ----
        if (level.staffAlert && plugin.configHolder.notifyAdmins) {
            val alert = MessageUtil.colorize(
                "&c[AntiLitematica] &e${player.name} &7被处理 (&f${level.action.displayName}&7): &f$msg &8[count=${record.count}]",
            )
            plugin.server.onlinePlayers
                .filter { it.hasPermission("antilitematica.notify") }
                .forEach { it.sendMessage(alert) }
        }

        // ---- 数据库记录 ----
        plugin.database.insertDetection(
            DetectionRecord(
                uuid = player.uniqueId,
                name = player.name,
                channel = channel,
                modDescription = "graduated:${level.action.name}",
                action = mapAction(level.action),
            ),
        )

        // ---- 出站通知（Discord / QQ） ----
        plugin.notificationService?.notifyAlert(
            "渐进惩罚",
            listOf(
                "玩家: ${player.name} (${player.uniqueId})",
                "动作: ${level.action.displayName}（第 ${record.count} 次 / 累计 ${record.totalViolations}）",
                "原因: ${level.reason}",
            ),
        )

        // ---- 反作弊联动 ----
        plugin.integrationManager.flag(
            player,
            "antilitematica:graduated",
            record.count,
            "level=${level.action.name}",
        )
    }

    /** 渐进动作 -> 检测记录动作（DB 兼容） */
    private fun mapAction(action: PunishmentAction): ActionType = when (action) {
        PunishmentAction.WARN -> ActionType.WARN
        PunishmentAction.KICK -> ActionType.KICK
        PunishmentAction.TEMPBAN, PunishmentAction.BAN -> ActionType.BAN
    }
}
